File: Java/MergeSort.java
package Java;

import java.io.*;
import java.util.Arrays;

public class MergeSort {

    public static int[] leerDatosDelArchivo(String nombreArchivo, int cantidad) {
        int[] datos = new int[cantidad];
        try (BufferedReader br = new BufferedReader(new FileReader(nombreArchivo))) {
            String linea;
            int i = 0;
            while ((linea = br.readLine()) != null && i < cantidad) {
                datos[i] = Integer.parseInt(linea.trim());  
                i++;
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
        return datos;  
    }

    public static int[] merge(int[] izquierda, int[] derecha) {
        int[] resultado = new int[izquierda.length + derecha.length];
        int i = 0, j = 0, k = 0;

        while (i < izquierda.length && j < derecha.length) {
            if (izquierda[i] < derecha[j]) {
                resultado[k] = izquierda[i];
                i++;
            } else {
                resultado[k] = derecha[j];
                j++;
            }
            k++;
        }

        while (i < izquierda.length) {
            resultado[k] = izquierda[i];
            i++;
            k++;
        }

        while (j < derecha.length) {
            resultado[k] = derecha[j];
            j++;
            k++;
        }

        return resultado; 
    }

    public static int[] mergeSort(int[] arr, int inicio, int fin) {
        if (fin - inicio <= 1) {
            return new int[]{arr[inicio]}; 
        }

        int mitad = (inicio + fin) / 2;
    
        int[] izquierda = mergeSort(arr, inicio, mitad);
    
        int[] derecha = mergeSort(arr, mitad, fin);

        // Aquí se ajusta para pasar el array 'derecha' con 'derecha-1' como un subarray
        return merge(izquierda, Arrays.copyOfRange(derecha, 0, derecha.length - 1)); // Excluimos el último elemento de derecha
    } 

    public static void main(String[] args) {
 
        String archivoDatos = "C:\\Users\\Marcelo Chiriboga\\Documentos\\Materias 5to Semestre - Software\\Análisis de Algoritmos\\Examen\\DatosCargados.txt";

        int cantidadDatosAProbar = 10000; 

        int[] numeros = leerDatosDelArchivo(archivoDatos, cantidadDatosAProbar);

        long inicioTemporizador = System.nanoTime();
        int[] listaOrdenada = mergeSort(numeros, 0 ,cantidadDatosAProbar);
        long finTemporizador = System.nanoTime();

        double tiempoTotalSegundos = (finTemporizador - inicioTemporizador) / 1_000_000_000.0;

        System.out.printf("Tiempo de ejecución para %d datos: %.3f segundos%n", cantidadDatosAProbar, tiempoTotalSegundos);

        if (cantidadDatosAProbar <= 50) {
            System.out.print("Lista ordenada: ");
            for (int num : listaOrdenada) {
                System.out.print( num + " | " );
            }
            System.out.println();
        }
    }
}

File: python/MergeSort.py
import time

def leer_datos_del_archivo(nombre_archivo, cantidad):
    with open(nombre_archivo, "r") as archivo:
        datos = [int(linea.strip()) for linea in archivo] 
    return datos[:cantidad]

def merge(lista_izquierda, lista_derecha):
    i = 0
    j = 0
    k = 0
    resultado = [0] * (len(lista_izquierda) + len(lista_derecha)) 

    while i < len(lista_izquierda) and j < len(lista_derecha):
        if lista_izquierda[i] < lista_derecha[j]:
            resultado[k] = lista_izquierda[i]
            i += 1
        else:
            resultado[k] = lista_derecha[j]
            j += 1
        k += 1
    
    while i < len(lista_izquierda):
        resultado[k] = lista_izquierda[i]
        i += 1
        k += 1

    while j < len(lista_derecha):
        resultado[k] = lista_derecha[j]
        j += 1
        k += 1
    
    return resultado

def merge_sort(arr):
    if len(arr) <= 1:
        return arr
    
    mitad = len(arr) // 2
    izquierda = merge_sort(arr[:mitad]) 
    derecha = merge_sort(arr[mitad:])   

    return merge(izquierda, derecha)


archivo_datos = "C:\\Users\\Marcelo Chiriboga\\Documentos\\Materias 5to Semestre - Software\\Análisis de Algoritmos\\Examen\\DatosCargados.txt"

cantidad_datos_a_probar = 10000  

numeros = leer_datos_del_archivo(archivo_datos, cantidad_datos_a_probar)

inicio_temporal = time.perf_counter()
lista_ordenada = merge_sort(numeros) 
fin_temporal = time.perf_counter()

tiempo_total_segundos = fin_temporal - inicio_temporal

print(f"Tiempo de ejecución para {cantidad_datos_a_probar} datos: {tiempo_total_segundos:.3f} segundos")

if cantidad_datos_a_probar <= 50:
    print("Lista ordenada:", lista_ordenada)
